File: 199 -  BinaryTreeRightSideView.py
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root == None: return []
        q = []
        q.append(root)
        result = []
        while q:
            size = len(q)
            for i in range(0, size):
                curr = q.pop(0)
                if i == size - 1:
                    result.append(curr.val)
                if curr.left != None:
                    q.append(curr.left)
                if curr.right != None:
                    q.append(curr.right)
        return result
    

    # T.C=> O(N) = since we traverse through every node
    # S.C=> O(N) = using stack
    # Approach => Add the every element in a queue by following breadth size.
    # Take the size of the queue everytime and just add all the last element in the stack from the queue.

File: 199-binaryTree_rightSideView.java
//Approach-1: Using the BFS
//Overall Time Complexity: O(N)
//Overall Space Complexity: O(N)
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if(root == null) return res;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        
        while(! q.isEmpty()){
            int size = q.size();
            
            for(int i = 0; i < size; i++){
                TreeNode node = q.poll();
                
                //Right Side View
                if(i == size - 1){
                    res.add(node.val);
                }
                
                if(node.left != null) q.add(node.left);
                if(node.right != null) q.add(node.right);
            }
        }
        
        return res;
    }
}


//Approach-2: Using the DFS
//Overall Time Complexity: O(N)
//Overall Space Complexity: O(H) - Height of the recursion stack
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        dfs(root, 0, res);
        return res;
    }
    
    public void dfs(TreeNode node, int level, List<Integer> res){
        //Base case
        if(node == null) return;
        
        //Functionality
        if(res.size() == level){
            res.add(node.val);
        }
        dfs(node.right, level + 1, res);
        dfs(node.left, level + 1, res);
        return;
    }
}


//Approach-3: Using the DFS - moving from left to right
//Overall Time Complexity: O(N)
//Overall Space Complexity: O(H) - Height of the recursion stack
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        dfs(root, 0, res);
        return res;
    }
    
    public void dfs(TreeNode node, int level, List<Integer> res){
        //Base case
        if(node == null) return;
        
        //Functionality
        if(res.size() == level){
            res.add(node.val);
        }else{
            res.set(level, node.val);
        }
        
        dfs(node.left, level + 1, res);
        dfs(node.right, level + 1, res);
        
        return;
    }
}